kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -357,16 +356,18 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
     }
 
     private fun NodeList.notifyCompletion(cause: Throwable?) =
-        notifyHandlers<JobNode>(this, cause)
+        notifyHandlers(this, cause) { true }
 
-    private inline fun <reified T: JobNode> notifyHandlers(list: NodeList, cause: Throwable?) {
+    private inline fun notifyHandlers(list: NodeList, cause: Throwable?, predicate: (JobNode) -> Boolean) {
         var exception: Throwable? = null
-        list.forEach<T> { node ->
-            try {
-                node.invoke(cause)
-            } catch (ex: Throwable) {
-                exception?.apply { addSuppressed(ex) } ?: run {
-                    exception =  CompletionHandlerException("Exception in completion handler $node for $this", ex)
+        list.forEach { node ->
+            if (node is JobNode && predicate(node)) {
```

Comment:
Can it ever be not an instance of `JobNode`? I would assume no, then worth adding a comment that this check is here only for the flow typing

Reply:
It can also be a "list is closed" token. We could abstract it a bit better by having `forEach` recognize and skip the service nodes, but given that we intend to remove the list altogether, I'm not sure that's worth doing.

Reply:
Though right now (without the DCSS removal), yes, the type bound is always satisfied.

Reply:
Works for me